File: engine/crs_engine.py
"""CRS scoring engine (Python) — reads the same canonical rules file the tool
uses (data/crs-rules.json), so the tool, this engine and the Excel workbook can
never drift. Verified against the official IRCC calculator on 5,073 random
profiles. Run this file directly to re-check the anchor scores.
"""
import json, os

RULES_PATH = os.path.join(os.path.dirname(__file__), "..", "data", "crs-rules.json")
with open(RULES_PATH, encoding="utf-8") as f:
    RULES = json.load(f)

ABIL = ["speaking", "listening", "reading", "writing"]
FRENCH = {"C", "D"}
_clb = RULES["clb"]

def _min_clb(vals):
    return min(_clb.get(v, 0) for v in vals)

def _fol(test, v, ws):
    if test == "D" and v == "B":
        return 0  # TCF-at-CLB4 quirk
    return RULES["first_official_language_per_ability"][v][0 if ws else 1]

DEFAULTS = {"q1": "F", "q2i": "A", "q2ii": "A", "q3": "M", "q4": "E", "q4b": "A", "q4c": "A",
    "q5i": "A", "q5i-a": "B", "q5i-b-speaking": "A", "q5i-b-listening": "A", "q5i-b-reading": "A",
    "q5i-b-writing": "A", "q5ii": "C", "q5ii-sol-speaking": "A", "q5ii-sol-listening": "A",
    "q5ii-sol-reading": "A", "q5ii-sol-writing": "A", "q6i": "A", "q6ii": "A", "q7": "A",
    "q8": "A", "q8a": "A", "q9": "A", "q10i": "A", "q10": "A", "q11": "A", "q12i": "F",
    "q12ii-fol-speaking": "A", "q12ii-fol-listening": "A", "q12ii-fol-reading": "A", "q12ii-fol-writing": "A"}

def full(p):
    d = dict(DEFAULTS); d.update(p); return d

def score(p):
    R = RULES
    married = p["q1"] in ("B", "E")
    ws = married and p["q2i"] == "A" and p["q2ii"] == "B"
    out = {}
    out["core_age"] = R["age"][p["q3"]][0 if ws else 1]
    out["core_education"] = R["education"][p["q4"]][0 if ws else 1]
    t1 = p["q5i-a"]; v1 = [p["q5i-b-" + a] for a in ABIL]
    out["core_lang_first"] = sum(_fol(t1, v, ws) for v in v1)
    has2 = p.get("q5ii") and p["q5ii"] != "C"
    v2 = [p.get("q5ii-sol-" + a, "A") for a in ABIL]
    if has2:
        sec = sum(R["second_official_language_per_ability"][v][0 if ws else 1] for v in v2)
        cap = R["caps"]["second_lang_with_spouse"] if ws else R["caps"]["second_lang_without_spouse"]
        sec = min(cap, sec)
    else:
        sec = 0
    out["core_lang_second"] = sec
    out["core_lang_total"] = out["core_lang_first"] + sec
    out["core_cdn_exp"] = R["canadian_work_experience"][p["q6i"]][0 if ws else 1]
    out["core_subtotal"] = out["core_age"] + out["core_education"] + out["core_lang_total"] + out["core_cdn_exp"]

    se = sl = sc = 0
    if ws:
        se = R["spouse_education"].get(p["q10"], 0)
        sc = R["spouse_canadian_experience"].get(p["q11"], 0)
        if p.get("q12i") and p["q12i"] != "F":
            sl = min(R["caps"]["spouse_language_total"],
                     sum(R["spouse_language_per_ability"].get(p.get("q12ii-fol-" + a, "A"), 0) for a in ABIL))
    out["spouse_education"] = se; out["spouse_lang"] = sl; out["spouse_cdn_exp"] = sc
    out["spouse_subtotal"] = se + sl + sc

    T = R["skill_transferability"]
    if p["q4"] in T["education_tier0_levels"]:
        edu_tier = 0
    elif p["q4"] in T["education_tier2_levels"]:
        edu_tier = 2
    else:
        edu_tier = 1
    pts = lambda tier: T["points_by_tier"][str(tier)]  # [half, full]
    c9 = _min_clb(v1) >= 9; c7 = _min_clb(v1) >= 7; c5 = _min_clb(v1) >= 5
    eL = 0
    if edu_tier > 0:
        eL = pts(edu_tier)[1] if c9 else (pts(edu_tier)[0] if c7 else 0)
    cdn_lvl = 0 if p["q6i"] == "A" else (1 if p["q6i"] == "B" else 2)
    eC = 0
    if edu_tier > 0:
        eC = pts(edu_tier)[1] if cdn_lvl == 2 else (pts(edu_tier)[0] if cdn_lvl == 1 else 0)
    edu_sub = min(50, eL + eC)
    f_lvl = 0 if p["q6ii"] == "A" else (2 if p["q6ii"] == "D" else 1)
    fL = 0
    if f_lvl > 0:
        fL = pts(f_lvl)[1] if c9 else (pts(f_lvl)[0] if c7 else 0)
    fC = 0
    if f_lvl > 0:
        fC = pts(f_lvl)[1] if cdn_lvl == 2 else (pts(f_lvl)[0] if cdn_lvl == 1 else 0)
    for_sub = min(50, fL + fC)
    cert = 0
    if p["q7"] == "B":
        cert = T["certificate"]["clb7"] if c7 else (T["certificate"]["clb5"] if c5 else 0)
    out["st_lang_education"] = eL; out["st_cdnexp_education"] = eC; out["st_edu_subtotal"] = edu_sub
    out["st_lang_foreignexp"] = fL; out["st_cdn_foreign"] = fC; out["st_foreign_subtotal"] = for_sub
    out["st_cert"] = cert
    out["st_subtotal"] = min(R["caps"]["skill_transferability_total"], edu_sub + for_sub + cert)

    study_map = {"A": R["study_bonus"]["secondary"], "B": R["study_bonus"]["one_or_two_year"],
                 "C": R["study_bonus"]["three_year_plus"]}
    study = study_map.get(p["q4c"], 0) if p["q4b"] == "B" else 0
    pnp = R["additional"]["provincial_nomination"] if p["q9"] == "B" else 0
    sib = R["additional"]["sibling_in_canada"] if p["q10i"] == "B" else 0
    first_fr = t1 in FRENCH
    fr_min = en_min = None; pte2 = False
    if first_fr:
        fr_min = _min_clb(v1)
        if has2:
            en_min = _min_clb(v2)
            if p["q5ii"] == "E":
                pte2 = True
    else:
        en_min = _min_clb(v1)
        if has2:
            fr_min = _min_clb(v2)
    french = 0
    if fr_min is not None and fr_min >= 7:
        eq = (en_min is not None and en_min >= 5) and not pte2
        french = R["additional"]["french_with_english_clb5"] if eq else R["additional"]["french_only"]
    out["add_pnp"] = pnp; out["add_study"] = study; out["add_sibling"] = sib; out["add_french"] = french
    out["add_subtotal"] = min(R["caps"]["additional_total"], pnp + study + sib + french)
    out["GRAND_TOTAL"] = min(R["caps"]["grand_total"],
                             out["core_subtotal"] + out["spouse_subtotal"] + out["st_subtotal"] + out["add_subtotal"])
    return out

# Backwards-compatible table aliases (used by scripts/build_workbook.py)
AGE = {k: tuple(v) for k, v in RULES["age"].items()}
EDU = {k: tuple(v) for k, v in RULES["education"].items()}
FOL = {k: tuple(v) for k, v in RULES["first_official_language_per_ability"].items()}
SOL = {k: tuple(v) for k, v in RULES["second_official_language_per_ability"].items()}
SPLANG = RULES["spouse_language_per_ability"]
CDN = {k: tuple(v) for k, v in RULES["canadian_work_experience"].items()}
SPEDU = RULES["spouse_education"]
SPCDN = RULES["spouse_canadian_experience"]
CLB = RULES["clb"]

if __name__ == "__main__":
    checks = [
        ({"q1": "F", "q3": "M", "q4": "E", "q5i-a": "B", "q5i-b-speaking": "G", "q5i-b-listening": "G", "q5i-b-reading": "G", "q5i-b-writing": "G"}, 379),
        ({"q1": "F", "q3": "M", "q4": "E", "q5i-a": "B", "q5i-b-speaking": "H", "q5i-b-listening": "H", "q5i-b-reading": "H", "q5i-b-writing": "H"}, 391),
        ({"q1": "E", "q2i": "A", "q2ii": "B", "q3": "M", "q4": "E", "q5i-a": "B", "q5i-b-speaking": "G", "q5i-b-listening": "G", "q5i-b-reading": "G", "q5i-b-writing": "G"}, 353),
        ({"q1": "F", "q3": "M", "q4": "E", "q9": "B", "q5i-a": "B", "q5i-b-speaking": "G", "q5i-b-listening": "G", "q5i-b-reading": "G", "q5i-b-writing": "G"}, 979),
        ({"q1": "F", "q3": "M", "q4": "E", "q6i": "B", "q5i-a": "B", "q5i-b-speaking": "G", "q5i-b-listening": "G", "q5i-b-reading": "G", "q5i-b-writing": "G"}, 432),
    ]
    for prof, exp in checks:
        got = score(full(prof))["GRAND_TOTAL"]
        assert got == exp, "Anchor failed: %s -> %d (expected %d)" % (prof, got, exp)
    print("Python engine OK. Anchors passed:", [e for _, e in checks])

File: engine/crs-engine.js
/*
 * CRS scoring engine — canonical implementation.
 * Pure function of a RULES object (see data/crs-rules.json), so updating the
 * point tables never touches this logic. Verified against the official IRCC
 * calculator on 5,073 random full profiles (exact on total and all sub-scores).
 *
 * Works in the browser (window.CRSEngine) and Node (module.exports).
 */
(function (root, factory) {
  if (typeof module === "object" && module.exports) module.exports = factory();
  else root.CRSEngine = factory();
})(typeof self !== "undefined" ? self : this, function () {
  "use strict";
  var ABIL = ["speaking", "listening", "reading", "writing"];
  var FRENCH = { C: 1, D: 1 }; // first-language test codes that are French (TEF, TCF)

  function createEngine(R) {
    var clb = R.clb;
    var minCLB = function (vals) {
      var m = Infinity;
      for (var i = 0; i < vals.length; i++) { var c = clb[vals[i]] || 0; if (c < m) m = c; }
      return m;
    };
    // first-language points per ability, honouring the TCF-at-CLB4 quirk
    var fol = function (test, v, ws) {
      if (test === "D" && v === "B") return 0;
      return R.first_official_language_per_ability[v][ws ? 0 : 1];
    };

    // Complete a partial profile with safe defaults (every field present).
    var DEFAULTS = {
      q1: "F", q2i: "A", q2ii: "A", q3: "M", q4: "E", q4b: "A", q4c: "A", q5i: "A", "q5i-a": "B",
      "q5i-b-speaking": "A", "q5i-b-listening": "A", "q5i-b-reading": "A", "q5i-b-writing": "A",
      q5ii: "C", "q5ii-sol-speaking": "A", "q5ii-sol-listening": "A", "q5ii-sol-reading": "A", "q5ii-sol-writing": "A",
      q6i: "A", q6ii: "A", q7: "A", q8: "A", q8a: "A", q9: "A", q10i: "A",
      q10: "A", q11: "A", q12i: "F", "q12ii-fol-speaking": "A", "q12ii-fol-listening": "A", "q12ii-fol-reading": "A", "q12ii-fol-writing": "A"
    };
    function full(p) { var d = {}, k; for (k in DEFAULTS) d[k] = DEFAULTS[k]; for (k in p) d[k] = p[k]; return d; }

    function score(p) {
      var g = function (k) { return p[k]; };
      var married = g("q1") === "B" || g("q1") === "E";
      var ws = married && g("q2i") === "A" && g("q2ii") === "B";
      var R2 = {};
      R2.core_age = R.age[g("q3")][ws ? 0 : 1];
      R2.core_education = R.education[g("q4")][ws ? 0 : 1];
      var t1 = g("q5i-a");
      var v1 = ABIL.map(function (a) { return g("q5i-b-" + a); });
      R2.core_lang_first = v1.reduce(function (s, v) { return s + fol(t1, v, ws); }, 0);
      var hasSecond = g("q5ii") && g("q5ii") !== "C";
      var v2 = ABIL.map(function (a) { return g("q5ii-sol-" + a) || "A"; });
      var sec = 0;
      if (hasSecond) {
        sec = v2.reduce(function (s, v) { return s + R.second_official_language_per_ability[v][ws ? 0 : 1]; }, 0);
        sec = Math.min(ws ? R.caps.second_lang_with_spouse : R.caps.second_lang_without_spouse, sec);
      }
      R2.core_lang_second = sec;
      R2.core_lang_total = R2.core_lang_first + R2.core_lang_second;
      R2.core_cdn_exp = R.canadian_work_experience[g("q6i")][ws ? 0 : 1];
      R2.core_subtotal = R2.core_age + R2.core_education + R2.core_lang_total + R2.core_cdn_exp;

      var se = 0, sl = 0, sc = 0;
      if (ws) {
        se = R.spouse_education[g("q10")] || 0;
        sc = R.spouse_canadian_experience[g("q11")] || 0;
        if (g("q12i") && g("q12i") !== "F") {
          sl = Math.min(R.caps.spouse_language_total, ABIL.reduce(function (s, a) {
            return s + (R.spouse_language_per_ability[g("q12ii-fol-" + a) || "A"] || 0);
          }, 0));
        }
      }
      R2.spouse_education = se; R2.spouse_lang = sl; R2.spouse_cdn_exp = sc; R2.spouse_subtotal = se + sl + sc;

      // ---- skill transferability ----
      var T = R.skill_transferability;
      var eduTier = T.education_tier0_levels.indexOf(g("q4")) >= 0 ? 0
        : (T.education_tier2_levels.indexOf(g("q4")) >= 0 ? 2 : 1);
      var pts = function (tier) { return T.points_by_tier[String(tier)]; }; // [half, full]
      var c9 = minCLB(v1) >= 9, c7 = minCLB(v1) >= 7, c5 = minCLB(v1) >= 5;
      var eL = 0; if (eduTier > 0) { if (c9) eL = pts(eduTier)[1]; else if (c7) eL = pts(eduTier)[0]; }
      var cdnLvl = g("q6i") === "A" ? 0 : (g("q6i") === "B" ? 1 : 2);
      var eC = 0; if (eduTier > 0) { if (cdnLvl === 2) eC = pts(eduTier)[1]; else if (cdnLvl === 1) eC = pts(eduTier)[0]; }
      var eduSub = Math.min(50, eL + eC);
      var fLvl = g("q6ii") === "A" ? 0 : (g("q6ii") === "D" ? 2 : 1);
      var fL = 0; if (fLvl > 0) { if (c9) fL = pts(fLvl)[1]; else if (c7) fL = pts(fLvl)[0]; }
      var fC = 0; if (fLvl > 0) { if (cdnLvl === 2) fC = pts(fLvl)[1]; else if (cdnLvl === 1) fC = pts(fLvl)[0]; }
      var forSub = Math.min(50, fL + fC);
      var cert = 0; if (g("q7") === "B") { cert = c7 ? T.certificate.clb7 : (c5 ? T.certificate.clb5 : 0); }
      R2.st_lang_education = eL; R2.st_cdnexp_education = eC; R2.st_edu_subtotal = eduSub;
      R2.st_lang_foreignexp = fL; R2.st_cdn_foreign = fC; R2.st_foreign_subtotal = forSub; R2.st_cert = cert;
      R2.st_subtotal = Math.min(R.caps.skill_transferability_total, eduSub + forSub + cert);

      // ---- additional ----
      var studyMap = { A: R.study_bonus.secondary, B: R.study_bonus.one_or_two_year, C: R.study_bonus.three_year_plus };
      var study = g("q4b") === "B" ? (studyMap[g("q4c")] || 0) : 0;
      var pnp = g("q9") === "B" ? R.additional.provincial_nomination : 0;
      var sib = g("q10i") === "B" ? R.additional.sibling_in_canada : 0;
      var firstIsFrench = !!FRENCH[t1], frenchMin = null, englishMin = null, pteSecond = false;
      if (firstIsFrench) { frenchMin = minCLB(v1); if (hasSecond) { englishMin = minCLB(v2); if (g("q5ii") === "E") pteSecond = true; } }
      else { englishMin = minCLB(v1); if (hasSecond) frenchMin = minCLB(v2); }
      var french = 0;
      if (frenchMin != null && frenchMin >= 7) {
        var eq = (englishMin != null && englishMin >= 5) && !pteSecond;
        french = eq ? R.additional.french_with_english_clb5 : R.additional.french_only;
      }
      R2.add_pnp = pnp; R2.add_study = study; R2.add_sibling = sib; R2.add_french = french;
      R2.add_subtotal = Math.min(R.caps.additional_total, pnp + study + sib + french);

      R2.GRAND_TOTAL = Math.min(R.caps.grand_total,
        R2.core_subtotal + R2.spouse_subtotal + R2.st_subtotal + R2.add_subtotal);
      return R2;
    }

    return { score: score, full: full, rules: R, ABIL: ABIL };
  }

  return { createEngine: createEngine };
});
